Introduce Qute templating extensions

mkouba opened the pull request:
- qute core as an independent project
- qute core extension
- qute resteasy extension
- first version of guide

NOTE: Extensions are considered `preview`.

mkouba committed:
Introduce Qute extensions

- qute core as an independent project
- qute core extension
- qute resteasy extension
- first version of guide

gastaldi reviewed (dismissed):
LGTM

gastaldi commented on docs/src/main/asciidoc/qute.adoc:
```diff
@@ -0,0 +1,271 @@
+    }
+}
+----
+<1> If there is no qualifier provided the field name is used to locate the template. In this particular case, we're injecting a template with path `META-INF/resources/templates/hello.txt`.
```
What qualifier can be provided to override the field name selection?

mkouba replied:
`@ResourcePath`. It's mentioned in the last example but it will be better to mention it here as well. Thanks for spotting this!

gastaldi commented on docs/src/main/asciidoc/qute.adoc:
```diff
@@ -0,0 +1,271 @@
+----
+<1> In this case, we use the `@ResourcePath` qualifier to specify the template path: `META-INF/resources/templates/reports/v1/report_01.html`.
+<2> Use the `@Scheduled` annotation to instruct Quarkus to execute this method on the half hour. For more information see the link:scheduler[Scheduler] guide.
+<3> The `Tempalte.getResult()` method triggers rendering. Note that this method blocks the current thread.
```
```suggestion
<3> The `Template.getResult()` method triggers rendering. Note that this method blocks the current thread.
```

gastaldi commented on docs/src/main/asciidoc/qute.adoc:
```diff
@@ -0,0 +1,271 @@
+</html>
+----
+<1> The loop section makes it possible to iterate over iterables, maps and streams.
+<2> This value expression is using elvis notation - if name is null the default value is used.
```
```suggestion
<2> This value expression is using elvis operator - if name is null the default value is used.
```

Maybe add a link to what Elvis https://en.wikipedia.org/wiki/Elvis_operator

gastaldi commented on docs/src/main/asciidoc/qute.adoc:
```diff
@@ -0,0 +1,271 @@
+</body>
+</html>
+----
+<1> Optional parameter declaration. Qute attempts to validate all expressions that reference paramater `item`.
```
How multiple parameters declarations are handled (should they be declared in another line or separated by a comma)?

mkouba replied:
Each param declaration is in a new tag. Anyway, this will be covered in the ref guide I'm currently working on. There's so many things and the guide should be brief and concise.

gsmet commented:
This PR is 134 files. Marking it as `waiting-for-ci` 3 hours after its opening is clearly not going to fly.

Please let more people review this stuff as it's a significant addition.

gsmet commented:
I'll have a look myself before the end of the week but I still have a couple of other PRs to review first.

mkouba commented:
> This PR is 134 files. 

Yes, and unfortunately it can't be reduced.

> Marking it as waiting-for-ci 3 hours after its opening is clearly not going to fly.

That's how @gastaldi works. You have to forgive him ;-). I think that nobody expected that this PR is merged immediately after the CI is green. However, 1) Most of the stuff was already discussed separately, 2) this is the first PR related to Qute but definitely not the last one - the extension is marked as preview and I'd like to merge the current state ASAP and improve what we have later because I see no benefits in discussing PRs for weeks.

gsmet commented:
> because I see no benefits in discussing PRs for weeks

It's not about discussing for weeks, it's about reviewing the code and the documentation. I'm not saying we should hold this PR to have more design discussions if it wasn't clear.

But we need a proper code review.

I'm just asking for others - including me - to have a chance at reviewing.

mkouba commented:
> I'm just asking for others - including me - to have a chance at reviewing.

No problem. And I think that nobody is against.

gsmet commented:
Btw, my remark on the size of the PR wasn't negative. I know it's a massive addition so no wonder it's a massive piece of code.

Just to clarify, adding `waiting-for-ci` basically means: please merge as soon as CI is green. So it should only be used in those specific cases.

gastaldi commented:
@mkouba is there a reason why `META-INF/resources/templates`  was chosen? Because the raw templates are publicly available in the webapp (eg. http://localhost:8080/templates/hello.txt)

gastaldi commented:
Is there any way to fetch the raw template contents from `Template`?

FroMage commented:
I'm not sure we _have_ to review the `Qute` core code. The extension, sure, but the rest is an independent project. It's as if we asked to review the resteasy code when we bump the resteasy dependency version number.

FroMage commented on docs/src/main/asciidoc/qute.adoc:
```diff
@@ -0,0 +1,271 @@
+public class HelloResource {
+
+    @Inject
+    Template hello; <1>
```
Didn't we have a variant where I didn't need to inject anything and it would default to the "class/method" template name?

mkouba replied:
Nope. It was part of your proposal but I was against it ;-).

FroMage replied:
Well, I'm still in favour of it, so can we have it or do I need to send a PR? Of course we can keep this variant, but I don't feel this scales if you have more than one method doing rendering.

mkouba replied:
I think that the right approach would be to 1) create an issue and 2) send a PR after this one gets merged. Thanks for understanding ;-)

FroMage replied:
Absolutely, which is why none of the issues I raised here should block this PR.

gastaldi commented:
The `qute` extension is never displayed in the list of loaded extensions: `Installed features: [cdi, resteasy]`

FroMage commented on docs/src/main/asciidoc/qute.adoc:
```diff
@@ -0,0 +1,271 @@
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    public Rendering get(@QueryParam("name") String name) {
```
I find `Rendering` to be a very confusing name. Perhaps `TemplateInstance` or `TemplateRendering`, but it's not really a "rendering instance". It is a template instance, with data. Perhaps the actual `Template` should be `TemplateFactory` and this `Rendering` should be `Template`. After all, in Java we never use `Foo` and `FooInstance`, we would call them `FooBuilder/Factory` and `Foo` for the instances.

mkouba replied:
It's not a template instance. It's "template instance + data + some config". So the best name would be probably `RenderingConfig` or `RenderingContext`. -1 for `TemplateFactory`. We can discuss the name later though.

gastaldi replied:
Not really useful, but in the reports module in Seam 3 we had `ReportDefinition` and `Report`: https://github.com/seam/reports/tree/develop/api/src/main/java/org/jboss/seam/reports

FroMage replied:
Well, to me, "template instance" can hold data. Like "Map" holds data.

mkouba replied:
Hm, `TemplateDefinition` sounds good. @FroMage WDYT?

@gastaldi `ReportDefinition` javadoc: _"A report definition object (normally called a "template")"_. ;-)

gastaldi replied:
@mkouba yeah, some Report engines call it that way :)

FroMage replied:
I'm totally fine with `TemplateDefinition` for the field, and `Template` for the object returned by the REST method.

mkouba replied:
Hm, I think I will try the `Template` -> `TemplateInstance` first. It sounds more natural to me. I'll update the PR and we can continue the discussion afterwards ;-).

FroMage commented on docs/src/main/asciidoc/qute.adoc:
```diff
@@ -0,0 +1,271 @@
+        return item.render().data("item", service.findItem(id)); <2>
+    }
+
+    @TemplateExtension <3>
```
I think we should advise people from the start to put their extensions in a separate file, by convention, even though they can put them anywhere. Something like `QuteExtensions.java` or `TemplateExtensions.java`.

mkouba replied:
Well, that's a matter of taste. For demos and guides it's definitely easier to declate the method where it's needed. Of course, we should introduce some good practices in the reference guide.  I can also imagine that that tooling would notify the developer that his application already contains a duplicit extension method for example.

FroMage replied:
For a demo, sure. But as soon as you have extension methods in two classes it becomes hard to remember the list and if you're not reimplementing one that already exists.

In fact, in Play 1, the framework had its own `JavaExtensions` class, and you had to extend it to add new methods. You didn't even need an annotation. Even if we don't go that way, we could allow the annotation on the class to avoid adding it on all its methods, if we advise for grouping.

mkouba replied:
> becomes hard to remember the list and if you're not reimplementing 

Yes, see my comment about tooling above.

> You didn't even need an annotation.

There is `TemplateExtension.matchName()` which could be useful. I'd stay with the current approach for now.

FroMage replied:
> Yes, see my comment about tooling above.

That's a way to generate errors on duplicates, but doesn't help you finding the set of extension methods. If we have a convention it becomes trivial: "Ctrl Shift r > WhateverExtension" and you have the full list ;)

> There is TemplateExtension.matchName() which could be useful. I'd stay with the current approach for now.

 Absolutely, but I mean if we advise to group them in a file, we could allow to place the annotation on the class and all its static methods would be registered even if they didn't have any method-annotation. I'm not saying we must remove the method annotation.

FroMage commented on docs/src/main/asciidoc/qute.adoc:
```diff
@@ -0,0 +1,271 @@
+
+    @Scheduled(cron="0 30 * * * ?") <2>
+    void generate() {
+        String result = report.render()
```
Similarly to what I said above, I don't feel like this method should be called `render` since it doesn't render.

mkouba replied:
Ok so what about `setup()`, `prepare()` or even `start()`?

gastaldi replied:
Just curious: Is a method here really necessary? Can't we just go with `report.data(..).render()`?

FroMage replied:
I think the name you're looking for is `instantiate` ;) We can probably find a fitting name once we find an agreement for `Rendering`, I'm sure it will follow.

FroMage replied:
> Just curious: Is a method here really necessary? Can't we just go with report.data(..).render()?

That was my first thought as well. You can start with the first parameter here and put the other ones on its result:

```java
String results = report.data("samples, service.get())
 .data("now", LocalDateTime.now())
 .render()
```

mkouba replied:
Yes, we could and it's probably a good idea. Although in theory, you don't always need to pass data and use e.g. `@Named` beans directly in the template.

FroMage replied:
Well we can have an overload that doesn't take data, provided we find its name ;)

FroMage commented on docs/src/main/asciidoc/qute.adoc:
```diff
@@ -0,0 +1,271 @@
+        String result = report.render()
+            .data("samples", service.get())
+            .data("now", java.time.LocalDateTime.now())
+            .getResult(); <3>
```
This method could be called `render()` though ;)

FroMage commented on extensions/qute-resteasy/deployment/src/test/java/io/quarkus/qute/resteasy/deployment/ItemResource.java:
```diff
@@ -0,0 +1,38 @@
+public class ItemResource {
+
+    @Inject
+    VariantTemplate item;
```
I'm not sure why we need to make variants explicit here. It requires to modify the controller when we could just add a new template file and it could be picked up as a variant. The fact that more than one version exists should be enough to guide the implementation to know there are variants.

FroMage commented on extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/DefaultTemplateExtensions.java:
```diff
@@ -0,0 +1,39 @@
+
+    @SuppressWarnings("rawtypes")
+    @TemplateExtension(matchName = ANY)
+    public static Object map(Map map, String name) {
```
So what's the rationale for deciding to put this in the extension rather than in Qute-core?

mkouba replied:
I don't remember ;-). I need to check the history.

FroMage replied:
Haha.

mkouba replied:
Hm, I think I remember now - the problem is that programatically added resolvers are ignored during validation of param declarations. So if there was `map.foo` it would fail even if there is a map entry for key `foo`. This area will definitely need more tests and improvements.

FroMage replied:
Ah, I guess probably the programmatic API doesn't have as much typing info as the annotation way? I thought the annoation would strictly use the programmatic API.

mkouba replied:
Well, most of the core APIs currently don't care about "type-safe validation" at all. It's an optional feature implemented on the quarkus integration level. Of course, we could try to change it later.

FroMage replied:
Ah, OK, I get it. In that case I suggest to move the validation up into Qute, even if you make it default to non-validated outside of Quarkus.

OR, perhaps a simpler trick would be to add just the typing info to the programmatic API, and make the Quarkus extension use this for checking?

FroMage reviewed (approved):
LGTM: all the remarks and questions I have can be solved later in new issues.

mkouba commented:
I've renamed `Rendering` to `TemplateInstance` and added and renamed a few methods in the API. Tomorrow, I'd like to move the base location to `src/main/resources/templates`.

mkouba commented:
It seems we have a problem with Windows and separators :-(. I'll try to address it tomorrow.

nimo23 commented:
Why can the template engine and syntax not be based on` jakarta.el.jar`?

mkouba committed:
Move templates under src/main/resources/templates

mkouba commented:
The failure of the windows build is caused by https://github.com/quarkusio/quarkus/issues/5821.

jaikiran committed:
Use filesystem specific name separator while trying to resolve a resource in RuntimeClassLoader

mkouba commented:
@jaikiran now I can see a weird error:
```
[ERROR] io.quarkus.qute.deployment.TypeSafeLoopTest  Time elapsed: 0 s  <<< ERROR!
java.nio.file.DirectoryNotEmptyException: D:\a\1\s\extensions\qute\deployment\target\quarkus-unit-test8157101450501518519
java.nio.file.DirectoryNotEmptyException: D:\a\1\s\extensions\qute\deployment\target\quarkus-unit-test8157101450501518519
	at sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:266)
	at sun.nio.fs.AbstractFileSystemProvider.delete(AbstractFileSystemProvider.java:103)
	at java.nio.file.Files.delete(Files.java:1126)
	at io.quarkus.test.QuarkusUnitTest$5.postVisitDirectory(QuarkusUnitTest.java:365)
	at io.quarkus.test.QuarkusUnitTest$5.postVisitDirectory(QuarkusUnitTest.java:345)
	at java.nio.file.Files.walkFileTree(Files.java:2688)
	at java.nio.file.Files.walkFileTree(Files.java:2742)
	at io.quarkus.test.QuarkusUnitTest.afterAll(QuarkusUnitTest.java:345)
```
AFAIK the directory should not be empty (see https://github.com/quarkusio/quarkus/blob/master/test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java#L345-L368). I'll try to run a windows box to see what's really happening.

mkouba committed:
QuteProcessor - fix validation of template injection points on windows

- also fix RuntimeClassLoader.createDefaultProtectionDomain(Path)

gsmet commented on independent-projects/qute/core/src/main/java/io/quarkus/qute/Expressions.java:
```diff
@@ -0,0 +1,107 @@
+        return value.indexOf(LEFT_BRACKET) != -1;
+    }
+
+    public static String parserVirtualMethodName(String value) {
```
```suggestion
    public static String parseVirtualMethodName(String value) {
```

gsmet commented on independent-projects/qute/core/src/main/java/io/quarkus/qute/Expressions.java:
```diff
@@ -0,0 +1,107 @@
+            String params = value.substring(start + 1, value.length() - 1);
+            return splitParts(params, Expressions::isParamSeparator);
+        }
+        throw new IllegalArgumentException("Not a virtual method: " + value);
```
Any reason why you have this check here and not in `parserVirtualMethodName`?

mkouba replied:
Sorry, I don't remember. Anyway, this code will need some more consolidation :-).

gsmet commented on independent-projects/qute/core/src/main/java/io/quarkus/qute/MemberKey.java:
```diff
@@ -0,0 +1,97 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Do we want copyright headers? I thought we ruled them out.

There are a few here and there.

mkouba replied:
I'll remove them.

gsmet commented on independent-projects/qute/core/src/main/java/io/quarkus/qute/ReflectionValueResolver.java:
```diff
@@ -0,0 +1,175 @@
+
+        Object base = context.getBase();
+        MemberKey key = MemberKey.newInstance(base, context.getName());
+        MemberWrapper wrapper = memberCache.computeIfAbsent(key, ReflectionValueResolver::findWrapper).orElse(null);
```
Sanne says that if you expect a lot of hits, it's more performant to do a proper `get` before that.

mkouba replied:
1) reflection resolver is just a fallback, 2) `get()` without sync would not be atomic

jaikiran commented:
@mkouba, just a quick input - keeping aside those failed to delete issues (in the `afterAll` test methods), I think the real failures probably boil down to this logic here https://github.com/mkouba/quarkus/blob/qute-next/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java#L502

So the `getPath` in `TemplateBuildItem` is a `Path` that will be using OS/filesystem specific separator. Then these filesystem specific paths are compared against "resource" paths (which always is "/" separated - for example `bar/bar.txt`) here https://github.com/mkouba/quarkus/blob/qute-next/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java#L547 which I think gets into issues on Windows OS. 

You know this code better than me, so ignore this input if it isn't right :) I had a few spare minutes and happened to find a Windows OS so just quickly ran the failing `InjectionTest` to see what might be the issue.

gsmet commented on independent-projects/qute/core/src/main/java/io/quarkus/qute/ValueResolvers.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    /**
+     * {@code foo.or(bar)},{@code foo or true},{@code name ?: 'elvis'}
```
do we also support a plain ternary operator?

mkouba replied:
Not now.

gsmet commented on pom.xml:
```diff
@@ -43,6 +43,7 @@
         <module>independent-projects/arc</module>
         <module>independent-projects/bootstrap</module>
         <module>independent-projects/tools</module>
+        <module>independent-projects/qute</module>
```
Note to myself: I will need to adjust the release scripts once this one is in.

gsmet commented on extensions/qute-resteasy/deployment/src/main/java/io/quarkus/qute/resteasy/deployment/QuteResteasyProcessor.java:
```diff
@@ -0,0 +1,23 @@
+
+    @BuildStep
+    ReflectiveHierarchyIgnoreWarningBuildItem ignoreReflectiveWarning() {
+        return new ReflectiveHierarchyIgnoreWarningBuildItem(DotName.createSimple(TemplateInstance.class.getName()));
```
I think it's more that we completely want to ignore it for RESTEasy? Because if for some reason the `TemplateInstance` class is in the index, you really don't want to include it in reflection.

In this case, for now, the type should be added there instead: https://github.com/quarkusio/quarkus/blob/master/extensions/resteasy-common/deployment/src/main/java/io/quarkus/resteasy/common/deployment/ResteasyDotNames.java#L69 .

We probably should have a better mechanism based on build items but that will do for now.

mkouba replied:
1) well, it shouldn't be in the index, 2) if it is, it's probably not a big problem, or?, 3) I'd like to avoid `ResteasyDotNames.TYPES_IGNORED_FOR_REFLECTION` if possible.

gsmet reviewed (changes requested):
I added some comments and questions.

There are a lot of very minor things but also a couple of more important interrogations.

mkouba commented on core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java:
```diff
@@ -46,6 +46,7 @@
     public static final String REACTIVE_MYSQL_CLIENT = "reactive-mysql-client";
     public static final String NEO4J = "neo4j";
     public static final String OIDC = "oidc";
+    public static final String QUTE = "qute";
```
A user will see `[qute, resteasy]`... I don't think there's a reason to add `qute-resteasy` or `resteasy-qute`.

gastaldi replied:
Out of curiosity: Is is possible to include `qute-resteasy` if  `qute` and `resteasy` are added to the project?

mkouba replied:
I don't think this could be done automatically ATM. But yes, it would make sense.

gsmet replied:
So... it's not that obvious to me.

If you have [qute, resteasy], that doesn't mean that you have the RESTEasy Qute support. The dependency could be missing and nothing will work.

So you could have two applications returning the exact same set of extensions and behaving differently. This is a problem IMHO.

And even for us or for support, when they will ask for the extension list, they will have a partial information.

gsmet replied:
Well, ignoring my comment doesn't magically resolve the issue ;)

I really think we should make it easy for support to know the list of installed extensions. As explained above, right now, they might have the same list of extensions and totally different behaviors.

RESTEasy Qute is a user exposed extension so it should be displayed here.

mkouba replied:
> Well, ignoring my comment doesn't magically resolve the issue ;)

That's a pity ;-). Well, I didn't know we have `resteasy-jsonb` and others. I will add it but it looks weird. Just my 2c.

mkouba commented:
@jaikiran Yep, you're right. I also added some logging and found out that `io.quarkus.runner.RuntimeClassLoader.createDefaultProtectionDomain(Path)` always fails on windows because the windows separator is quoted: 
```
java.net.URISyntaxException: Relative path in absolute URI: file:D:%5Ca%5C1%5Cs%5Cextensions%5Cmailer%5Cdeployment%5Ctarget%5Cquarkus-unit-test6389559223281079023
```
I'll fix the `QuteProcessor` first and we'll see if those `DirectoryNotEmptyException`s disappear ;-).

mkouba commented on docs/src/main/asciidoc/qute.adoc:
```diff
@@ -0,0 +1,271 @@
+<title>Report {now}</title>
+</head>
+<body>
+    <h1>Report {now}</h1>
```
Not ATM. We do `toString()`. But it's extensible and we could add special formatting helpers later.

mkouba commented on extensions/qute-resteasy/runtime/src/main/java/io/quarkus/qute/resteasy/TemplateResponseFilter.java:
```diff
@@ -0,0 +1,80 @@
+                    mediaType = null;
+                }
+            } else {
+                // TODO how to get media type from non-variant templates?
```
That's on the plan. Remember that this is the first PR incorporating working prototype.

mkouba commented on extensions/qute-resteasy/deployment/pom.xml:
```diff
@@ -0,0 +1,58 @@
+    <name>Quarkus - Qute RESTEasy - Deployment</name>
+
+    <dependencies>
+        <dependency>
```
I don't understand. We don't need the runtime modules in this deployment module and for a user app the runtime deps will be added automatically.

mkouba commented on extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/IncorrectExpressionBuildItem.java:
```diff
@@ -0,0 +1,21 @@
+    public final int line;
+    public final String templateId;
+
+    public IncorrectExpressionBuildItem(String expression, String property, String clazz, int line, String templateId) {
```
I wish we had a proper javadoc on at least 80% of build items ;-).

mkouba commented on extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/ImplicitValueResolverBuildItem.java:
```diff
@@ -0,0 +1,19 @@
+
+public final class ImplicitValueResolverBuildItem extends MultiBuildItem {
+
+    private final ClassInfo clazz;
```
> I think we tried to avoid that.

Why? I'm pretty sure we do propagate jandex elements in many build items and since they're immutable it should be fine.

mkouba commented on extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java:
```diff
@@ -0,0 +1,837 @@
+        for (IncorrectExpressionBuildItem incorrectExpression : incorrectExpressions) {
+            if (incorrectExpression.clazz != null) {
+                errors.add(String.format(
+                        "Incorrect expression: %s\n\t- property [%s] not found on class [%s] nor handled by an extension method\n\t- found in template [%s] on line %s",
```
The error message ends with `on line %s`. But I agree that the message format is TBD.

mkouba commented on extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java:
```diff
@@ -0,0 +1,837 @@
+import org.jboss.jandex.ParameterizedType;
+import org.jboss.jandex.Type;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Ah, that was an oversight... we have jboss-logging in the qute core as well.

mkouba commented on extensions/qute/runtime/pom.xml:
```diff
@@ -0,0 +1,55 @@
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus.qute</groupId>
+            <artifactId>qute-rxjava</artifactId>
```
AFAIK at least SR Fault Tolerance and SR Context Propagation depend on it too.

FroMage replied:
I don't think CP depends on it. FT does.

mkouba commented on extensions/qute/runtime/src/main/java/io/quarkus/qute/api/ResourcePath.java:
```diff
@@ -0,0 +1,34 @@
+public @interface ResourcePath {
+
+    /**
+     * @return the path relative from {@code META-INF/resources/${basePath}}, must not be an empty string
```
This is just an outdated javadoc. The templates are located in `src/main/resources/templates`.

mkouba commented on extensions/qute/runtime/src/main/java/io/quarkus/qute/api/VariantTemplate.java:
```diff
@@ -0,0 +1,21 @@
+package io.quarkus.qute.api;
```
Well, this is the same situation we have with ArC - `io.quarkus.qute` is used in the core artifact and in java9+ it's not good to use the same package in multiple artifacts. Maybe we could use some "subpackage"?

mkouba commented on extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/DefaultTemplateExtensions.java:
```diff
@@ -0,0 +1,39 @@
+                return map.containsKey(name);
+            default:
+                return Result.NOT_FOUND;
+        }
```
We have it for `Map` because we support dot notation to access map values: `{map.myKey}`. And this would break type checking because `myKey` could be just anything. This is one of the things we'll try to improve ;-).

mkouba commented on extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/EngineProducer.java:
```diff
@@ -0,0 +1,172 @@
+     */
+    private Optional<Reader> locate(String path) {
+        // Use the system separator
+        // TODO String path = path.replace('/', File.separatorChar);
```
Yep, outdated comment ;-).

mkouba commented on extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/QuteConfig.java:
```diff
@@ -0,0 +1,18 @@
+    /**
+     * The set of suffixes used when attempting to locate a template.
+     */
+    @ConfigItem(defaultValue = "html,txt")
```
-1 from me. Templating engine is not a good tool to build JSON/XML files.

mkouba commented on extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java:
```diff
@@ -0,0 +1,837 @@
+                    AnnotationValue propertiesValue = templateData.value("properties");
+                    if (propertiesValue == null || propertiesValue.equals(v.value("properties"))) {
+                        propertiesValue = AnnotationValue.createBooleanValue("properties", false);
+                    }
```
TBH I don't remember the details and don't have the capacity to process it right now ;-). I can remove the comment but I don't think it would help anything...

mkouba commented on extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/EngineProducer.java:
```diff
@@ -0,0 +1,172 @@
+            in = locatePath(templatePath);
+        }
+        if (in != null) {
+            return Optional.of(new InputStreamReader(in, Charset.forName("utf-8")));
```
Yep, it should be safe. We'll document the contract that the engine is responsible for closing the returned value after use.

mkouba commented on extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/QuteRecorder.java:
```diff
@@ -0,0 +1,63 @@
+        if (engine != null) {
+            engine.removeTemplates(id -> {
+                // Exact match or path starts with id, e.g. "items.html" starts with "items"
+                boolean remove = pathIds.contains(id) || pathIds.stream().anyMatch(pid -> pid.startsWith(id));
```
I'm going to remove this method completely. The original use case was to drop changed templates during dev mode instead of restart but this is no more possible due to type checking.

mkouba commented on extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/VariantTemplateProducer.java:
```diff
@@ -0,0 +1,200 @@
+
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Yep, the problem is that `slf4j-api` comes from `quarkus-core` and so it's always on the class path.

mkouba commented on independent-projects/qute/core/pom.xml:
```diff
@@ -0,0 +1,32 @@
+        </dependency>
+        <dependency>
+            <groupId>org.reactivestreams</groupId>
+            <artifactId>reactive-streams</artifactId>
```
I don't think so - it's part of the API, see `io.quarkus.qute.TemplateInstance`. Anyway IIRC, it was recommended from Clement and Stephane to include this one.

FroMage replied:
Yeah, this one is the 4 interface jar. It's harmless and required by RESTEasy core already, and many other libs.

mkouba committed:
Rename quarkus-qute-resteasy to quarkus-resteasy-qute

- plus other fixes

mkouba commented:
@jaikiran it turns out that the `DirectoryNotEmptyException` was caused by the fact that we did not explicitly close the stream produced by `java.nio.file.Files.list()`.

@gsmet this PR should be ready now.

I'm working on a next PR with reference documentation and few minor updates.

jaikiran commented:
>@jaikiran it turns out that the DirectoryNotEmptyException was caused by the fact that we did not explicitly close the stream produced by java.nio.file.Files.list().

Good to know :)

gsmet commented on core/deployment/src/main/java/io/quarkus/runner/RuntimeClassLoader.java:
```diff
@@ -522,14 +526,22 @@ private ProtectionDomain createDefaultProtectionDomain(Path applicationClasspath
+                //    path = path.replace(File.separatorChar, '/');
+                //}
+                //URI uri = new URI("file", null, path, null);
+                URI uri = applicationClasspath.toUri();
```
Is it normal that we have all this commented code here (and one line below)?

mkouba replied:
No. It's a leftover from a fix I did and wasn't quite sure if it's correct. I'll remove the commented code.

gsmet commented on extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/EngineProducer.java:
```diff
@@ -0,0 +1,169 @@
+import javax.inject.Singleton;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
I still see some SL4J logging. Shouldn't it all use JBoss Logging?

mkouba replied:
I agree. I'll search the code again. And again, it's not very useful that SLF4J is always on the class path :-(.

gsmet commented on extensions/resteasy-qute/deployment/src/main/java/io/quarkus/resteasy/qute/deployment/QuteResteasyProcessor.java:
```diff
@@ -0,0 +1,23 @@
+import io.quarkus.resteasy.common.spi.ResteasyJaxrsProviderBuildItem;
+import io.quarkus.resteasy.qute.runtime.TemplateResponseFilter;
+
+public class QuteResteasyProcessor {
```
Minor but should be `ResteasyQuteProcessor` to be consistent.

gsmet commented on extensions/resteasy-qute/pom.xml:
```diff
@@ -0,0 +1,23 @@
+    </parent>
+
+    <artifactId>quarkus-resteasy-qute-parent</artifactId>
+    <name>Quarkus - RESTEasy Qute</name>
```
Minor but should be `Quarkus - RESTEasy - Qute` to follow the convention we have for Jackson and JSON-B. I can do it in a follow-up if you don't feel like doing it.

Same for the runtime and deployment artifacts obviously.

mkouba replied:
What's the convention? The dash in the name?

mkouba replied:
Ok, it's the dash. I misread your comment. Ok, done.

gsmet reviewed (changes requested):
@mkouba we are nearly there I think.

Could you have a look at my latest comments? It would be great if we could merge it tomorrow.

BTW, I'm wondering if it would be better to squash the resteasy-qute renaming commit? But maybe it was conflicting? If so don't bother.

gsmet reviewed (approved):
OK, let's get this monster in :).

@mkouba thanks for your work and your patience :).

mkouba commented:
The `JPAFunctionalityInGraalITCase` fails with:
```
2019-12-02 19:05:45,066 ERROR [org.hib.eng.jdb.spi.SqlExceptionHelper] (main) No timezone mapping entry for 'Etc/GMT'
javax.persistence.PersistenceException: [PersistenceUnit: templatePU] Unable to build Hibernate SessionFactory
	at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.persistenceException(FastBootEntityManagerFactoryBuilder.java:113)
	at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:67)
	at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:54)
	at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:80)
	at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:55)
	at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:109)
	at io.quarkus.hibernate.orm.runtime.JPAConfig.startAll(JPAConfig.java:57)
	at io.quarkus.hibernate.orm.runtime.HibernateOrmRecorder.startAllPersistenceUnits(HibernateOrmRecorder.java:82)
	at io.quarkus.deployment.steps.HibernateOrmProcessor$startPersistenceUnits182.deploy_0(HibernateOrmProcessor$startPersistenceUnits182.zig:70)
	at io.quarkus.deployment.steps.HibernateOrmProcessor$startPersistenceUnits182.deploy(HibernateOrmProcessor$startPersistenceUnits182.zig:36)
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:134)
	at io.quarkus.runtime.Application.start(Application.java:87)
	at io.quarkus.runtime.Application.run(Application.java:210)
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:41)
Caused by: org.hibernate.exception.GenericJDBCException: Unable to open JDBC Connection for DDL execution
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:47)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
	at org.hibernate.resource.transaction.backend.jta.internal.DdlTransactionIsolatorJtaImpl.<init>(DdlTransactionIsolatorJtaImpl.java:62)
	at org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorBuilderImpl.buildDdlTransactionIsolator(JtaTransactionCoordinatorBuilderImpl.java:46)
	at org.hibernate.tool.schema.internal.HibernateSchemaManagementTool.getDdlTransactionIsolator(HibernateSchemaManagementTool.java:175)
	at org.hibernate.tool.schema.internal.HibernateSchemaManagementTool.buildGenerationTargets(HibernateSchemaManagementTool.java:135)
	at org.hibernate.tool.schema.internal.SchemaDropperImpl.doDrop(SchemaDropperImpl.java:110)
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.performDatabaseAction(SchemaManagementToolCoordinator.java:145)
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.process(SchemaManagementToolCoordinator.java:73)
	at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:320)
	at org.hibernate.boot.internal.SessionFactoryBuilderImpl.build(SessionFactoryBuilderImpl.java:462)
	at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:65)
	... 12 more
Caused by: java.sql.SQLException: No timezone mapping entry for 'Etc/GMT'
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:129)
```
It's very likely unrelated. I'll try to re-run the test once again.

gsmet commented:
Yes, it's unrelated. Looks like CI cache issue all over again.
